File: apps/web/src/lib/analysis.ts
import type {
  CoverageUnit,
  DistributionMetrics,
  TextQualityMetrics,
} from "@/lib/evaluation";
import { authenticatedFetch } from "@/lib/browser-auth";

export interface PageResult<T> {
  readonly items: T[];
  readonly total: number;
  readonly offset: number;
  readonly limit: number;
}

export interface PhonologyStatus {
  readonly cache_available: boolean;
  readonly loaded: boolean;
  readonly revision: string;
  readonly sha256: string;
  readonly stats: null | {
    readonly inventory_count: number;
    readonly language_count: number;
    readonly segment_count: number;
  };
}

export interface LanguageSummary {
  readonly iso639_3: string;
  readonly glottocode: string;
  readonly language_name: string;
  readonly inventory_count: number;
  readonly sources: string[];
}

export interface EspeakMapping {
  readonly espeak_code: string;
  readonly iso639_3: string;
}

export interface InventorySources {
  readonly identifier: string;
  readonly sources: string[];
}

export interface FeatureCatalog {
  readonly names: string[];
}

export interface PhoneticFeature {
  readonly name: string;
  readonly value: string;
}

export interface Segment {
  readonly phoneme: string;
  readonly segment_class: "consonant" | "vowel" | "tone";
  readonly marginal: boolean;
  readonly allophones: string[];
  readonly features: PhoneticFeature[];
  readonly glyph_id: string;
}

export interface Inventory {
  readonly inventory_id: number;
  readonly language_name: string;
  readonly iso639_3: string;
  readonly glottocode: string;
  readonly specific_dialect: string | null;
  readonly source: string;
  readonly segments: Segment[];
  readonly phonemes: string[];
  readonly consonants: string[];
  readonly vowels: string[];
  readonly tones: string[];
  readonly marginal_phonemes: string[];
  readonly size: number;
  readonly consonant_count: number;
  readonly vowel_count: number;
  readonly tone_count: number;
}

export interface UnitCountInput {
  readonly unit: string;
  readonly count: number;
}

export interface UnitWeightInput {
  readonly unit: string;
  readonly weight: number;
}

export type RateValue =
  | { readonly status: "finite"; readonly value: number }
  | {
      readonly status: "positive_infinity" | "not_computed";
      readonly value: null;
    };

export interface ErrorRateDetail {
  readonly index: number;
  readonly reference: string;
  readonly hypothesis: string;
  readonly wer: RateValue;
  readonly cer: RateValue;
  readonly per: RateValue;
}

export interface ErrorRateResult {
  readonly wer: RateValue;
  readonly cer: RateValue;
  readonly per: RateValue;
  readonly ser: RateValue;
  readonly details: ErrorRateDetail[];
}

export interface CoverageSnapshot {
  readonly sentence_index: number;
  readonly coverage: number;
  readonly covered_count: number;
  readonly new_units_count: number;
  readonly new_units: string[];
}

export interface CoverageTrajectory {
  readonly unit: CoverageUnit;
  readonly target_size: number;
  readonly coverages: number[];
  readonly gains: number[];
  readonly snapshots: CoverageSnapshot[];
}

export class AnalysisContractError extends Error {
  constructor() {
    super("The analysis API returned an incompatible response.");
    this.name = "AnalysisContractError";
  }
}

export class AnalysisApiError extends Error {
  readonly status: number;
  readonly code: string;
  readonly requestId: string | null;

  constructor(
    status: number,
    code: string,
    message: string,
    requestId: string | null = null,
  ) {
    super(message);
    this.name = "AnalysisApiError";
    this.status = status;
    this.code = code;
    this.requestId = requestId;
  }
}

function record(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function finite(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function nonnegativeInteger(value: unknown): value is number {
  return Number.isInteger(value) && (value as number) >= 0;
}

function strings(value: unknown): value is string[] {
  return (
    Array.isArray(value) && value.every((item) => typeof item === "string")
  );
}

function nullableFinite(value: unknown): value is number | null {
  return value === null || finite(value);
}

export function parsePage<T>(
  value: unknown,
  itemParser: (item: unknown) => T,
): PageResult<T> {
  if (!record(value) || !Array.isArray(value.items))
    throw new AnalysisContractError();
  const items = value.items.map(itemParser);
  if (
    !nonnegativeInteger(value.total) ||
    !nonnegativeInteger(value.offset) ||
    !nonnegativeInteger(value.limit)
  ) {
    throw new AnalysisContractError();
  }
  return {
    items,
    total: value.total,
    offset: value.offset,
    limit: value.limit,
  };
}

export function parsePhonologyStatus(value: unknown): PhonologyStatus {
  if (
    !record(value) ||
    typeof value.cache_available !== "boolean" ||
    typeof value.loaded !== "boolean" ||
    typeof value.revision !== "string" ||
    !/^[0-9a-f]{40}$/u.test(value.revision) ||
    typeof value.sha256 !== "string" ||
    !/^[0-9a-f]{64}$/u.test(value.sha256)
  ) {
    throw new AnalysisContractError();
  }
  if (value.stats !== null) {
    if (
      !record(value.stats) ||
      !nonnegativeInteger(value.stats.inventory_count) ||
      !nonnegativeInteger(value.stats.language_count) ||
      !nonnegativeInteger(value.stats.segment_count)
    ) {
      throw new AnalysisContractError();
    }
  }
  return value as unknown as PhonologyStatus;
}

export function parseLanguage(value: unknown): LanguageSummary {
  if (
    !record(value) ||
    typeof value.iso639_3 !== "string" ||
    typeof value.glottocode !== "string" ||
    typeof value.language_name !== "string" ||
    !nonnegativeInteger(value.inventory_count) ||
    !strings(value.sources)
  ) {
    throw new AnalysisContractError();
  }
  return value as unknown as LanguageSummary;
}

export function parseMapping(value: unknown): EspeakMapping {
  if (
    !record(value) ||
    typeof value.espeak_code !== "string" ||
    typeof value.iso639_3 !== "string"
  ) {
    throw new AnalysisContractError();
  }
  return value as unknown as EspeakMapping;
}

export function parseSegment(value: unknown): Segment {
  if (
    !record(value) ||
    typeof value.phoneme !== "string" ||
    (value.segment_class !== "consonant" &&
      value.segment_class !== "vowel" &&
      value.segment_class !== "tone") ||
    typeof value.marginal !== "boolean" ||
    !strings(value.allophones) ||
    !Array.isArray(value.features) ||
    typeof value.glyph_id !== "string"
  ) {
    throw new AnalysisContractError();
  }
  const features = value.features.map((feature) => {
    if (
      !record(feature) ||
      typeof feature.name !== "string" ||
      typeof feature.value !== "string" ||
      !/^[+\-0](?:,[+\-0])*$/u.test(feature.value) ||
      feature.value.length > 15
    ) {
      throw new AnalysisContractError();
    }
    return feature as unknown as PhoneticFeature;
  });
  return { ...(value as unknown as Omit<Segment, "features">), features };
}

export function parseInventory(value: unknown): Inventory {
  if (
    !record(value) ||
    !nonnegativeInteger(value.inventory_id) ||
    typeof value.language_name !== "string" ||
    typeof value.iso639_3 !== "string" ||
    typeof value.glottocode !== "string" ||
    (value.specific_dialect !== null &&
      typeof value.specific_dialect !== "string") ||
    typeof value.source !== "string" ||
    !Array.isArray(value.segments) ||
    !strings(value.phonemes) ||
    !strings(value.consonants) ||
    !strings(value.vowels) ||
    !strings(value.tones) ||
    !strings(value.marginal_phonemes) ||
    !nonnegativeInteger(value.size) ||
    !nonnegativeInteger(value.consonant_count) ||
    !nonnegativeInteger(value.vowel_count) ||
    !nonnegativeInteger(value.tone_count)
  ) {
    throw new AnalysisContractError();
  }
  const segments = value.segments.map(parseSegment);
  return { ...(value as unknown as Omit<Inventory, "segments">), segments };
}

const distributionFloatFields = [
  "entropy",
  "normalized_entropy",
  "jsd_uniform",
  "coefficient_of_variation",
  "count_ratio",
  "pcd_uniform",
] as const;
const distributionIntegerFields = [
  "min_count",
  "max_count",
  "zero_count",
] as const;

export function parseDistribution(value: unknown): DistributionMetrics {
  if (
    !record(value) ||
    !distributionFloatFields.every((field) => finite(value[field])) ||
    !distributionIntegerFields.every((field) =>
      nonnegativeInteger(value[field]),
    ) ||
    !nullableFinite(value.jsd_reference) ||
    !nullableFinite(value.pearson_correlation)
  ) {
    throw new AnalysisContractError();
  }
  return value as unknown as DistributionMetrics;
}

const qualityFloatFields = [
  "sentence_length_words_mean",
  "sentence_length_words_median",
  "sentence_length_words_std",
  "sentence_length_phonemes_mean",
  "sentence_length_phonemes_median",
  "sentence_length_phonemes_std",
  "type_token_ratio",
  "hapax_ratio",
] as const;
const qualityIntegerFields = [
  "sentence_length_words_min",
  "sentence_length_words_max",
  "sentence_length_phonemes_min",
  "sentence_length_phonemes_max",
  "total_words",
  "unique_words",
] as const;

export function parseTextQuality(value: unknown): TextQualityMetrics {
  if (
    !record(value) ||
    !qualityFloatFields.every((field) => finite(value[field])) ||
    !qualityIntegerFields.every((field) => nonnegativeInteger(value[field])) ||
    !nullableFinite(value.flesch_reading_ease) ||
    !nullableFinite(value.flesch_kincaid_grade)
  ) {
    throw new AnalysisContractError();
  }
  return value as unknown as TextQualityMetrics;
}

function parseRate(value: unknown): RateValue {
  if (!record(value) || typeof value.status !== "string")
    throw new AnalysisContractError();
  if (value.status === "finite" && finite(value.value))
    return { status: "finite", value: value.value };
  if (
    (value.status === "positive_infinity" || value.status === "not_computed") &&
    value.value === null
  ) {
    return { status: value.status, value: null };
  }
  throw new AnalysisContractError();
}

export function parseErrorRates(value: unknown): ErrorRateResult {
  if (!record(value) || !Array.isArray(value.details))
    throw new AnalysisContractError();
  const details = value.details.map((detail) => {
    if (
      !record(detail) ||
      !nonnegativeInteger(detail.index) ||
      typeof detail.reference !== "string" ||
      typeof detail.hypothesis !== "string"
    ) {
      throw new AnalysisContractError();
    }
    return {
      index: detail.index,
      reference: detail.reference,
      hypothesis: detail.hypothesis,
      wer: parseRate(detail.wer),
      cer: parseRate(detail.cer),
      per: parseRate(detail.per),
    };
  });
  return {
    wer: parseRate(value.wer),
    cer: parseRate(value.cer),
    per: parseRate(value.per),
    ser: parseRate(value.ser),
    details,
  };
}

export function parseTrajectory(value: unknown): CoverageTrajectory {
  if (
    !record(value) ||
    (value.unit !== "phoneme" &&
      value.unit !== "diphone" &&
      value.unit !== "triphone") ||
    !nonnegativeInteger(value.target_size) ||
    !Array.isArray(value.coverages) ||
    !value.coverages.every((item) => finite(item) && item >= 0 && item <= 1) ||
    !Array.isArray(value.gains) ||
    !value.gains.every(nonnegativeInteger) ||
    !Array.isArray(value.snapshots)
  ) {
    throw new AnalysisContractError();
  }
  const snapshots = value.snapshots.map((snapshot) => {
    if (
      !record(snapshot) ||
      !nonnegativeInteger(snapshot.sentence_index) ||
      !finite(snapshot.coverage) ||
      snapshot.coverage < 0 ||
      snapshot.coverage > 1 ||
      !nonnegativeInteger(snapshot.covered_count) ||
      !nonnegativeInteger(snapshot.new_units_count) ||
      !strings(snapshot.new_units)
    ) {
      throw new AnalysisContractError();
    }
    return snapshot as unknown as CoverageSnapshot;
  });
  return {
    unit: value.unit,
    target_size: value.target_size,
    coverages: value.coverages as number[],
    gains: value.gains as number[],
    snapshots,
  };
}

async function api<T>(
  path: string,
  parser: (value: unknown) => T,
  init?: RequestInit,
): Promise<T> {
  const response = await authenticatedFetch(path, {
    cache: "no-store",
    headers: {
      Accept: "application/json",
      ...(init?.body ? { "Content-Type": "application/json" } : {}),
    },
    ...init,
  });
  if (!response.ok) {
    let body: unknown;
    try {
      body = await response.json();
    } catch {
      body = null;
    }
    if (record(body) && typeof body.message === "string") {
      throw new AnalysisApiError(
        response.status,
        typeof body.code === "string" ? body.code : "request_failed",
        body.message,
        typeof body.request_id === "string" ? body.request_id : null,
      );
    }
    throw new AnalysisApiError(
      response.status,
      "request_failed",
      `Request failed with HTTP ${response.status}.`,
    );
  }
  return parser((await response.json()) as unknown);
}

function query(
  parameters: Record<string, string | number | boolean | undefined>,
): string {
  const result = new URLSearchParams();
  for (const [key, value] of Object.entries(parameters)) {
    if (value !== undefined && value !== "") result.set(key, String(value));
  }
  const encoded = result.toString();
  return encoded ? `?${encoded}` : "";
}

export const analysisApi = {
  phonologyStatus: (signal?: AbortSignal) =>
    api("/api/v1/phonology/status", parsePhonologyStatus, { signal }),
  loadPhonology: (signal?: AbortSignal) =>
    api("/api/v1/phonology/load", parsePhonologyStatus, {
      method: "POST",
      signal,
    }),
  languages: (search: string, signal?: AbortSignal, offset = 0, limit = 50) =>
    api(
      `/api/v1/phonology/languages${query({ query: search, offset, limit })}`,
      (value) => parsePage(value, parseLanguage),
      { signal },
    ),
  espeakMappings: (
    search: string,
    signal?: AbortSignal,
    offset = 0,
    limit = 50,
  ) =>
    api(
      `/api/v1/phonology/espeak-mappings${query({ query: search, offset, limit })}`,
      (value) => parsePage(value, parseMapping),
      { signal },
    ),
  inventory: (
    identifier: string,
    source: string | undefined,
    union: boolean,
    signal?: AbortSignal,
  ) =>
    api(
      `/api/v1/phonology/inventories/${encodeURIComponent(identifier)}${query({ source: union ? undefined : source, union })}`,
      parseInventory,
      { signal },
    ),
  featureCatalog: (signal?: AbortSignal) =>
    api(
      "/api/v1/phonology/features",
      (value): FeatureCatalog => {
        if (!record(value) || !strings(value.names)) {
          throw new AnalysisContractError();
        }
        return { names: value.names };
      },
      { signal },
    ),
  inventorySources: (identifier: string, signal?: AbortSignal) =>
    api(
      `/api/v1/phonology/inventories/${encodeURIComponent(identifier)}/sources`,
      (value): InventorySources => {
        if (
          !record(value) ||
          typeof value.identifier !== "string" ||
          !strings(value.sources)
        ) {
          throw new AnalysisContractError();
        }
        return { identifier: value.identifier, sources: value.sources };
      },
      { signal },
    ),
  allInventories: (
    identifier: string,
    offset = 0,
    limit = 5,
    signal?: AbortSignal,
  ) =>
    api(
      `/api/v1/phonology/inventories/${encodeURIComponent(identifier)}/all${query({ offset, limit })}`,
      (value) => parsePage(value, parseInventory),
      { signal },
    ),
  segments: (
    identifier: string,
    filters: {
      readonly source?: string;
      readonly union?: boolean;
      readonly segmentClass?: "consonant" | "vowel" | "tone";
      readonly marginal?: boolean;
      readonly featureName?: string;
      readonly featureValue?: string;
      readonly offset?: number;
      readonly limit?: number;
    },
    signal?: AbortSignal,
  ) =>
    api(
      `/api/v1/phonology/inventories/${encodeURIComponent(identifier)}/segments${query(
        {
          source: filters.union ? undefined : filters.source,
          union: filters.union,
          segment_class: filters.segmentClass,
          marginal: filters.marginal,
          feature_name: filters.featureName,
          feature_value: filters.featureValue,
          offset: filters.offset ?? 0,
          limit: filters.limit ?? 50,
        },
      )}`,
      (value) => parsePage(value, parseSegment),
      { signal },
    ),
  distribution: (
    body: {
      counts: UnitCountInput[];
      target_units: string[];
      reference_distribution: UnitWeightInput[] | null;
    },
    signal?: AbortSignal,
  ) =>
    api("/api/v1/analyses/distribution", parseDistribution, {
      method: "POST",
      body: JSON.stringify(body),
      signal,
    }),
  textQuality: (
    body: { sentences: string[]; phoneme_sequences: string[][] },
    signal?: AbortSignal,
  ) =>
    api("/api/v1/analyses/text-quality", parseTextQuality, {
      method: "POST",
      body: JSON.stringify(body),
      signal,
    }),
  errorRates: (
    body: {
      references: string[];
      hypotheses: string[];
      reference_phonemes: string[][] | null;
      hypothesis_phonemes: string[][] | null;
      case_sensitive: boolean;
    },
    signal?: AbortSignal,
  ) =>
    api("/api/v1/analyses/error-rates", parseErrorRates, {
      method: "POST",
      body: JSON.stringify(body),
      signal,
    }),
  trajectory: (
    body: {
      phoneme_sequences: string[][];
      target_units: string[];
      unit: CoverageUnit;
    },
    signal?: AbortSignal,
  ) =>
    api("/api/v1/analyses/coverage-trajectory", parseTrajectory, {
      method: "POST",
      body: JSON.stringify(body),
      signal,
    }),
};

File: src/corpuskit/adapters/corpusgen/analysis.py
"""Typed CorpusGen boundary for deterministic corpus analyses."""

from __future__ import annotations

import math
from collections.abc import Callable
from typing import Protocol, cast

from pydantic import ValidationError

from corpuskit.domain import (
    CoverageSnapshot,
    CoverageTrajectory,
    CoverageTrajectoryRequest,
    DistributionAnalysisRequest,
    DistributionMetrics,
    EngineContractError,
    EngineUnavailableError,
    ErrorRatesAnalysis,
    ErrorRatesAnalysisRequest,
    InvalidRequestError,
    RateStatus,
    RateValue,
    SentenceErrorRate,
    TextQualityAnalysisRequest,
    TextQualityMetrics,
)


class DistributionLike(Protocol):
    entropy: float
    normalized_entropy: float
    jsd_uniform: float
    coefficient_of_variation: float
    min_count: int
    max_count: int
    count_ratio: float
    zero_count: int
    pcd_uniform: float
    jsd_reference: float | None
    pearson_correlation: float | None


class TextQualityLike(Protocol):
    sentence_length_words_mean: float
    sentence_length_words_median: float
    sentence_length_words_std: float
    sentence_length_words_min: int
    sentence_length_words_max: int
    sentence_length_phonemes_mean: float
    sentence_length_phonemes_median: float
    sentence_length_phonemes_std: float
    sentence_length_phonemes_min: int
    sentence_length_phonemes_max: int
    total_words: int
    unique_words: int
    type_token_ratio: float
    hapax_ratio: float
    flesch_reading_ease: float | None
    flesch_kincaid_grade: float | None


class SentenceErrorLike(Protocol):
    index: int
    reference: str
    hypothesis: str
    wer: float
    cer: float
    per: float | None


class ErrorRatesLike(Protocol):
    wer: float
    cer: float
    per: float | None
    ser: float
    details: list[SentenceErrorLike]


class CoverageSnapshotLike(Protocol):
    sentence_index: int
    coverage: float
    covered_count: int
    new_units_count: int
    new_units: list[str]


class CoverageTrajectoryLike(Protocol):
    snapshots: list[CoverageSnapshotLike]
    unit: str
    target_size: int


class DistributionComputer(Protocol):
    def __call__(
        self,
        phoneme_counts: dict[str, int],
        target_phonemes: list[str],
        reference_distribution: dict[str, float] | None = None,
    ) -> DistributionLike: ...


class TextQualityComputer(Protocol):
    def __call__(
        self,
        sentences: list[str],
        phoneme_sequences: list[list[str]],
    ) -> TextQualityLike: ...


class ErrorRatesComputer(Protocol):
    def __call__(
        self,
        references: list[str],
        hypotheses: list[str],
        reference_phonemes: list[list[str]] | None = None,
        hypothesis_phonemes: list[list[str]] | None = None,
        case_sensitive: bool = False,
    ) -> ErrorRatesLike: ...


class TrajectoryComputer(Protocol):
    def __call__(
        self,
        phoneme_sequences: list[list[str]],
        target_units: set[str],
        unit: str = "phoneme",
    ) -> CoverageTrajectoryLike: ...


def _default_distribution(
    counts: dict[str, int],
    targets: list[str],
    reference_distribution: dict[str, float] | None = None,
) -> DistributionLike:
    from corpusgen.evaluate.distribution import compute_distribution_metrics

    return cast(
        DistributionLike,
        compute_distribution_metrics(counts, targets, reference_distribution),
    )


def _default_text_quality(
    sentences: list[str],
    phoneme_sequences: list[list[str]],
) -> TextQualityLike:
    from corpusgen.evaluate.text_quality import compute_text_quality_metrics

    return cast(TextQualityLike, compute_text_quality_metrics(sentences, phoneme_sequences))


def _default_error_rates(
    references: list[str],
    hypotheses: list[str],
    reference_phonemes: list[list[str]] | None = None,
    hypothesis_phonemes: list[list[str]] | None = None,
    case_sensitive: bool = False,
) -> ErrorRatesLike:
    from corpusgen.evaluate.error_rates import compute_error_rates

    return cast(
        ErrorRatesLike,
        compute_error_rates(
            references,
            hypotheses,
            reference_phonemes=reference_phonemes,
            hypothesis_phonemes=hypothesis_phonemes,
            case_sensitive=case_sensitive,
        ),
    )


def _default_trajectory(
    phoneme_sequences: list[list[str]],
    target_units: set[str],
    unit: str = "phoneme",
) -> CoverageTrajectoryLike:
    from corpusgen.evaluate.trajectory import compute_coverage_trajectory

    return cast(
        CoverageTrajectoryLike,
        compute_coverage_trajectory(phoneme_sequences, target_units, unit),
    )


class CorpusgenAnalysisAdapter:
    """Normalize deterministic CorpusGen metrics into JSON-safe DTOs."""

    def __init__(
        self,
        *,
        distribution_computer: DistributionComputer | None = None,
        text_quality_computer: TextQualityComputer | None = None,
        error_rates_computer: ErrorRatesComputer | None = None,
        trajectory_computer: TrajectoryComputer | None = None,
    ) -> None:
        self._distribution = distribution_computer or _default_distribution
        self._text_quality = text_quality_computer or _default_text_quality
        self._error_rates = error_rates_computer or _default_error_rates
        self._trajectory = trajectory_computer or _default_trajectory

    def distribution(self, request: DistributionAnalysisRequest) -> DistributionMetrics:
        operation = "analysis.distribution"
        reference = (
            {item.unit: item.weight for item in request.reference_distribution}
            if request.reference_distribution is not None
            else None
        )
        result = self._invoke(
            lambda: self._distribution(
                {item.unit: item.count for item in request.counts},
                list(request.target_units),
                reference,
            ),
            operation,
        )
        try:
            return DistributionMetrics(
                entropy=result.entropy,
                normalized_entropy=result.normalized_entropy,
                jsd_uniform=result.jsd_uniform,
                coefficient_of_variation=result.coefficient_of_variation,
                min_count=result.min_count,
                max_count=result.max_count,
                count_ratio=result.count_ratio,
                zero_count=result.zero_count,
                pcd_uniform=result.pcd_uniform,
                jsd_reference=result.jsd_reference,
                pearson_correlation=result.pearson_correlation,
            )
        except (AttributeError, TypeError, ValidationError, ValueError):
            raise EngineContractError(operation) from None

    def text_quality(self, request: TextQualityAnalysisRequest) -> TextQualityMetrics:
        operation = "analysis.text_quality"
        result = self._invoke(
            lambda: self._text_quality(
                list(request.sentences),
                [list(sequence) for sequence in request.phoneme_sequences],
            ),
            operation,
        )
        try:
            return TextQualityMetrics(
                sentence_length_words_mean=result.sentence_length_words_mean,
                sentence_length_words_median=result.sentence_length_words_median,
                sentence_length_words_std=result.sentence_length_words_std,
                sentence_length_words_min=result.sentence_length_words_min,
                sentence_length_words_max=result.sentence_length_words_max,
                sentence_length_phonemes_mean=result.sentence_length_phonemes_mean,
                sentence_length_phonemes_median=result.sentence_length_phonemes_median,
                sentence_length_phonemes_std=result.sentence_length_phonemes_std,
                sentence_length_phonemes_min=result.sentence_length_phonemes_min,
                sentence_length_phonemes_max=result.sentence_length_phonemes_max,
                total_words=result.total_words,
                unique_words=result.unique_words,
                type_token_ratio=result.type_token_ratio,
                hapax_ratio=result.hapax_ratio,
                flesch_reading_ease=result.flesch_reading_ease,
                flesch_kincaid_grade=result.flesch_kincaid_grade,
            )
        except (AttributeError, TypeError, ValidationError, ValueError):
            raise EngineContractError(operation) from None

    def error_rates(self, request: ErrorRatesAnalysisRequest) -> ErrorRatesAnalysis:
        operation = "analysis.error_rates"
        result = self._invoke(
            lambda: self._error_rates(
                list(request.references),
                list(request.hypotheses),
                (
                    [list(sequence) for sequence in request.reference_phonemes]
                    if request.reference_phonemes is not None
                    else None
                ),
                (
                    [list(sequence) for sequence in request.hypothesis_phonemes]
                    if request.hypothesis_phonemes is not None
                    else None
                ),
                request.case_sensitive,
            ),
            operation,
        )
        try:
            details = tuple(
                SentenceErrorRate(
                    index=item.index,
                    reference=item.reference,
                    hypothesis=item.hypothesis,
                    wer=self._rate(item.wer),
                    cer=self._rate(item.cer),
                    per=self._rate(item.per),
                )
                for item in result.details
            )
            if len(details) != len(request.references) or any(
                item.index != index for index, item in enumerate(details)
            ):
                raise ValueError("engine returned inconsistent error details")
            return ErrorRatesAnalysis(
                wer=self._rate(result.wer),
                cer=self._rate(result.cer),
                per=self._rate(result.per),
                ser=self._rate(result.ser),
                details=details,
            )
        except (AttributeError, TypeError, ValidationError, ValueError):
            raise EngineContractError(operation) from None

    def trajectory(self, request: CoverageTrajectoryRequest) -> CoverageTrajectory:
        operation = "analysis.coverage_trajectory"
        result = self._invoke(
            lambda: self._trajectory(
                [list(sequence) for sequence in request.phoneme_sequences],
                set(request.target_units),
                request.unit.value,
            ),
            operation,
        )
        try:
            if result.unit != request.unit.value or result.target_size != len(request.target_units):
                raise ValueError("engine returned a different trajectory contract")
            snapshots = tuple(
                CoverageSnapshot(
                    sentence_index=item.sentence_index,
                    coverage=item.coverage,
                    covered_count=item.covered_count,
                    new_units_count=item.new_units_count,
                    new_units=tuple(item.new_units),
                )
                for item in result.snapshots
            )
            if len(snapshots) != len(request.phoneme_sequences) or any(
                item.sentence_index != index
                or item.new_units_count != len(item.new_units)
                or item.covered_count > result.target_size
                for index, item in enumerate(snapshots)
            ):
                raise ValueError("engine returned inconsistent trajectory snapshots")
            return CoverageTrajectory(
                unit=request.unit,
                target_size=result.target_size,
                coverages=tuple(item.coverage for item in snapshots),
                gains=tuple(item.new_units_count for item in snapshots),
                snapshots=snapshots,
            )
        except (AttributeError, TypeError, ValidationError, ValueError):
            raise EngineContractError(operation) from None

    @staticmethod
    def _rate(value: float | None) -> RateValue:
        if value is None:
            return RateValue(status=RateStatus.NOT_COMPUTED, value=None)
        if math.isfinite(value):
            return RateValue(status=RateStatus.FINITE, value=value)
        if math.isinf(value) and value > 0:
            return RateValue(status=RateStatus.POSITIVE_INFINITY, value=None)
        raise ValueError("engine returned an invalid rate")

    @staticmethod
    def _invoke[T](call: Callable[[], T], operation: str) -> T:
        try:
            return call()
        except ValueError:
            raise InvalidRequestError(operation) from None
        except (RuntimeError, OSError):
            raise EngineUnavailableError(operation) from None
        except Exception:
            raise EngineUnavailableError(operation) from None


__all__ = ["CorpusgenAnalysisAdapter"]
